core-util/v2/detail/capture.hpp

```diff
@@ -0,0 +1,145 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FUNCTIONAL_DETAIL_CAPTURE_HPP
+#define FUNCTIONAL_DETAIL_CAPTURE_HPP
+
+#include <type_traits>
+#include <tuple>
+
+#include "interface.hpp"
+#include "allocators.hpp"
+
+namespace functional {
+namespace detail {
+
+
+namespace index {
+template<std::size_t ...>
+struct sequence { };
+
+template<std::size_t N, std::size_t ...S>
+struct generator : generator<N-1, N-1, S...> { };
+
+template<std::size_t ...S>
+struct generator<0, S...> {
+    typedef sequence<S...> type;
+};
+} // namespace index
+
+template <typename FunctionType, ContainerAllocator & Allocator, typename... CapturedTypes>
+class CaptureFirst;
+
+template <typename ReturnType, typename... ArgTypes, ContainerAllocator & Allocator, typename... CapturedTypes>
+class CaptureFirst <ReturnType(ArgTypes...), Allocator, CapturedTypes...>
+        : public FunctionInterface <ReturnType(ArgTypes...)> {
+public:
+    CaptureFirst(const std::tuple<CapturedTypes...>& t, Function<ReturnType(CapturedTypes..., ArgTypes...)>& f):
+        f(f), storage(t)
+    {}
+
+    virtual ReturnType operator () (ArgTypes&&... Args) {
```

Comment:
Sorry, I didn't saw that CaptureFirst inherit from FunctionInterface, it makes sense if it is virtual.